### PR TITLE
Add TreeNode, Tree, and related Visitor classes together with test files

### PR DESCRIPTION
**Pull request by serdarkacka**
Package `jtdiff.util` contains `TreeNode`, `Tree`, and related `Visitor` classes.
Package `jtdiff.util.tests` contains unit tests testing `TreeNode` and `Tree` classes. JUnit 4 is required to run the unit tests.

### CODE REVIEW

File: jtdiff/util/tests/TestTree.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		TreeNode eNode = new TreeNode("E");
+		aNode.addChild(bNode);
+		aNode.addChild(cNode);
+		cNode.addChild(dNode);
+		dNode.addChild(eNode);
+		treeThree = new Tree(aNode);
+		treeThree.buildCaches();
+	}
+}
+
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestTree.java]: Remove the unnecessary whitespace.

Comment (cagdasgerede) [inline jtdiff/util/tests/TestTree.java]: [minor] Remove the extra whitespace

File: jtdiff/util/tests/TestTree.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
+package jtdiff.util.tests;
+
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestTree.java:2]: [minor] Remove extra whitespace

File: jtdiff/util/tests/TestTree.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	public void setUp() throws Exception {
+		TreeNode aNode = new TreeNode("A");
+		TreeNode bNode = new TreeNode("B");
+		aNode.addChild(bNode);
+		TreeNode cNode = new TreeNode("C");
+		bNode.addChild(cNode);		
+		treeOne = new Tree(aNode);
+		treeOne.buildCaches();
+		
+
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestTree.java]: [minor] remove extra newline

File: jtdiff/util/tests/TestPreorderTraversal.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.junit.Test;
+
+/**
+ * 
+ * Tests <code>performPreorderTraversal</code> method of <code>Tree</code> class
+ *
+ */
+public class TestPreorderTraversal extends TestTree {
+
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestPreorderTraversal.java]: remove extra whitespace

File: jtdiff/util/tests/TestFatherOf.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	 */
+	@Test
+	public void testSuccess() {
+		assertEquals(3, treeTwo.fatherOf(4).preorderPosition());
+		assertEquals(1, treeTwo.fatherOf(3).preorderPosition());
+		assertEquals(1, treeTwo.fatherOf(2).preorderPosition());	
+	}
+	
+	/**
+	 * Father of root should be None
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestFatherOf.java]: None => null

File: jtdiff/util/Tree.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	public TreeNode fatherOf(int preorderPosition) throws IllegalArgumentException {
+		TreeNode node = this.preorderPositionToNode.get(preorderPosition);
+		if (node != null) {
+			return node.father();
+		}
+		
+		throw new IllegalArgumentException("No node at the given position: " + preorderPosition);
+	}
+	
+
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java]: remove extra newlines

File: jtdiff/util/Tree.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+			}
+			else {
+				break;
+			}
+		}
+		
+		return positions.iterator();
+	}
+	
+
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java]: remove extra newline

File: jtdiff/util/Tree.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+		while (fatherNode.preorderPosition() != parentPosition) {
+			currentChildNode = fatherNode;
+			fatherNode = currentChildNode.father();
+			if (fatherNode == null) {
+				return null;
+			}
+		}
+		return currentChildNode;
+	}
+
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java]: remove newlines

File: jtdiff/util/TreeNode.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	/**
+	 * Adds a new child to the node
+	 * @param node is a <code>TreeNode</code> object
+	 */
+	public void addChild(TreeNode node) {
+		this.children.add(node);
+		node.setFather(this);
+	}
+	
+
```
Comment (cagdasgerede) [inline jtdiff/util/TreeNode.java]: remove extra newlines

File: jtdiff/util/TreeNode.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	
+	/**
+	 * 
+	 * @return a <code>String</code> representation for this node for debugging
+	 */
+	public String debugString() {
+		return String.format("label: %s, preorder_position: %d", label, preorderPosition);
+	}
+
```
Comment (cagdasgerede) [inline jtdiff/util/TreeNode.java]: remove newlines

File: jtdiff/util/Tree.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	 * @param parentPosition
+	 * @param descendantPosition
+	 * @return the child of a node which is on the path from a 
+	 * descendant of the node to the node (there can only be one such child).
+	 * Returns null if no such node.
+	 * @throws IllegalArgumentException
+	 */
+	public TreeNode childOnPathFromDescendant(int parentPosition, int descendantPosition) 
+			throws IllegalArgumentException {
+
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java]: remove newline

File: jtdiff/util/Tree.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+	 * @throws IllegalArgumentException
+	 */
+	public TreeNode childOnPathFromDescendant(int parentPosition, int descendantPosition) 
+			throws IllegalArgumentException {
+		
+		TreeNode currentChildNode = this.nodeAt(descendantPosition);
+		TreeNode fatherNode = currentChildNode.father();
+		
+		if (fatherNode == null) {
+			throw new IllegalArgumentException("No father node at the given position: " + descendantPosition);
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java]: lines should be <=80 characters

File: jtdiff/util/Tree.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+	
+	
+	/**
+	 * Finds a child node between the parent and the descendant.
+	 * @param parentPosition
+	 * @param descendantPosition
+	 * @return the child of a node which is on the path from a 
+	 * descendant of the node to the node (there can only be one such child).
+	 * Returns null if no such node.
+	 * @throws IllegalArgumentException
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java:124]: IllegalArgumentException if no node found at the descendantPosition

File: jtdiff/util/Tree.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+	/**
+	 * Builds the cached preorder positions of the nodes in the tree.
+	 * Call this method after the tree structure is finalized
+	 */
+	public void buildCaches() {
+		Visitor visitor = new PreOrderMarkingVisitor(this);
+		this.performPreorderTraversal(visitor);
+	}
+
+
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java]: remove extra newline

File: jtdiff/util/Tree.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+	
+	/**
+	 * Performs a preorder traversal on the tree
+	 * @param visitor is used for taking an action on the visited node (the visitor must have a visit method)
+	 */
+	public void performPreorderTraversal(Visitor visitor) {
+		this.root.preorderTraversal(visitor);
+		
+	}
+
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java]: remove extra newline

File: jtdiff/util/Tree.java [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+	
+	/**
+	 * Does a preorder traversal and prints the node labels
+	 */
+	public void printPreorderTraversal() {
+		Visitor visitor = new DebugVisitor();
+		this.performPreorderTraversal(visitor);
+	}
+	
+
```
Comment (cagdasgerede) [inline jtdiff/util/Tree.java]: remove extra newline

File: jtdiff/util/TreeNode.java [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+	
+	/**
+	 * Returns an iterator for iterating TreeNode children 
+	 * @return Iterator in  <code>TreeNode</code> type
+	 */
+	public Iterator<TreeNode> children() {
+		return this.children.iterator();
+	}
+	
+
```
Comment (cagdasgerede) [inline jtdiff/util/TreeNode.java]: remove newline

File: jtdiff/util/tests/TestAncestorIterator.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.junit.Test;
+
+/**
+ * 
+ * Tests <code>ancestorIterator</code> method of <code>Tree</code> class
+ *
+ */
+public class TestAncestorIterator extends TestTree {
+
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestAncestorIterator.java]: remove newline

File: jtdiff/util/tests/TestNodeAt.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+		assertEquals("C", treeOne.nodeAt(3).label());
+		assertNull(treeOne.nodeAt(4));
+		
+		// Tree 2
+		assertEquals("A", treeTwo.nodeAt(1).label());
+		assertEquals("B", treeTwo.nodeAt(2).label());
+		assertEquals("C", treeTwo.nodeAt(3).label());
+		assertEquals("D", treeTwo.nodeAt(4).label());
+		assertNull(treeTwo.nodeAt(5));
+
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestNodeAt.java]: remove newline

File: jtdiff/util/tests/TestNodeAt.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+ * Tests <code>nodeAt</code> method of <code>Tree</code> class
+ *
+ */
+public class TestNodeAt extends TestTree {
+	/**
+	 * Testing getting a node from the tree by its preorder position
+	 */
+	@Test
+	public void testNodeAt() {
+		// Testing getting a node from the tree by its preorder position
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestNodeAt.java]: remove this comment and extra newlines

File: jtdiff/util/tests/TestFatherOf.java [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+		
+	}
+	
+	/**
+	 * Raise error when father of a non-existing node requested
+	 */
+	@Test(expected = IllegalArgumentException.class)
+	public void testWhenNoSuchNode() {
+		treeTwo.fatherOf(5);
+
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestFatherOf.java]: remove newline

File: jtdiff/util/tests/TestPreorderTraversal.java [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+	
+	/**
+	 * Successfully produce the list of visited node labels
+	 */
+	@Test
+	public void testSucess() {
+		TraversalVisitor visitor = new TraversalVisitor();
+		treeOne.performPreorderTraversal(visitor);
+		
+		String[] expected = new String[] {"label: A, preorder_position: 1",
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestPreorderTraversal.java]: move "label: A.... to the next line to align with other 2 "label... lines.

File: jtdiff/util/tests/TestPreorderTraversal.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+	            "label: B, preorder_position: 2",
+	            "label: C, preorder_position: 3"};
+		assertArrayEquals(expected, visitor.getTraversal());
+		
+		visitor = new TraversalVisitor();
+		treeTwo.performPreorderTraversal(visitor);
+		expected = new String[] {"label: A, preorder_position: 1",
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestPreorderTraversal.java]: ditto

File: jtdiff/util/tests/TestTree.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+	@Override
+	public void visit(TreeNode node) {
+		traversal.add(node.debugString());
+		
+	}
+	
+	public String[] getTraversal() {
+		return traversal.toArray(new String[traversal.size()]);
+	}
+
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestTree.java]: remove newline

File: jtdiff/util/tests/TestTree.java [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.ArrayList;
+
+import org.junit.Before;
+
+
+import jtdiff.util.Tree;
+import jtdiff.util.TreeNode;
+import jtdiff.util.Visitor;
+
+class TraversalVisitor extends Visitor {
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestTree.java:12]: Add (small) comment to describe what this class is

File: jtdiff/util/tests/TestTree.java [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+	public String[] getTraversal() {
+		return traversal.toArray(new String[traversal.size()]);
+	}
+	
+}
+
+
+/**
+ * 
+ * Tests the functionality of Tree
```
Comment (cagdasgerede) [inline jtdiff/util/tests/TestTree.java]: This class does not test anything now. It is used to build trees to be used by its children

**Review by cagdasgerede — CHANGES_REQUESTED** [review]
Minor comments (mostly formatting related)

**Comment by serdarkacka** [comment]
The issues you are referring are from commit  7e8ed9e. After that, I've submitted two more commits: fefe37f and  a5a29d6. The '>= 80 character' problems and some other style problems are solved in these commits.

**Review by cagdasgerede — APPROVED** [review]

**Pull request merged**